Feature/openapi automatic

provokateurin opened the pull request:
Closes https://github.com/provokateurin/nextcloud-neon/issues/1
Closes https://github.com/provokateurin/nextcloud-neon/issues/417
Depends on https://github.com/provokateurin/nextcloud-neon/pull/462 https://github.com/provokateurin/nextcloud-neon/pull/463 https://github.com/provokateurin/nextcloud-neon/pull/464 and https://github.com/nextcloud/server/pull/39223

Leptopoda commented:
Looks ok at the first glance.
I'll thoroughly review it in the coming days.

Leptopoda commented:
I might also rebase/redo my changes for #420 so we can land the automatic branch without any big changes afterwards (might be better for devs already using this lib to adjust only once).

provokateurin commented:
Sounds look a good plan, let's do it 👍

Leptopoda commented:
The ContentString serialization is fixed with #501 but the same test now fails with a 500 server error.
Let me know if this is also related to built_value

provokateurin commented:
I'll look into the test failure once your PR is merged

provokateurin commented:
I assume the problem is that we don't properly deserialize the parameters. OpenAPI can describe many ways to serialize values (https://swagger.io/docs/specification/serialization/) but we just assume something and then spec also doesn't say what it actually needs.

provokateurin commented:
Yep, server log confirms it: `Argument #5 ($shareTypes) must be of type array, string given`.

Leptopoda commented:
> The autocomplete test is still failing because we are not properly serializing the parameter.

But this isn't a problem with dynamite? 

> Yep, server log confirms it: Argument #5 ($shareTypes) must be of type array, string given.

The spec says it's a string (ContentString to be exact). After changing this the AutocompleteResult.status appears to also be wrong.

https://github.com/provokateurin/nextcloud-neon/pull/465/files#diff-24dc2feb72ef0ff5bc5c6e6b4fe8f5efd9693883efc2893bd0d1afd9feea73a3R1576-R1592

provokateurin commented:
I fixed the specs and when https://github.com/provokateurin/nextcloud-neon/pull/522 is merged it will work correctly. The autocomplete tests is still broken because of a bug that was I fixed but that is only in 28+.

Leptopoda commented:
We can still skip it until we switch to 28

provokateurin commented:
Shall we wait for the #420 work before merging this? The result should be a lot cleaner.

Leptopoda commented on packages/nextcloud/test/provisioning_api_test.dart:
```diff
@@ -14,24 +14,42 @@ void main() {
+        expect(apps.ocs.data.apps, hasLength(39));
+      });
+
+      test('Get app infos', () async {
+        for (final app in (await client.provisioningApi.apps.getApps()).ocs.data.apps) {
+          await client.provisioningApi.apps.getAppInfo(app: app);
+        }
+      });
```
I'm not a fan of tests without any expect (I can see why it's done though)

But I really dislike `(await client.provisioningApi.apps.getApps()).ocs.data.apps`
The result of getApps should be in a separate variable raising thje question why this isn't merged with the test above it.

Leptopoda commented on packages/nextcloud/test/core_test.dart:
```diff
@@ -69,8 +69,10 @@ void main() {
     });
 
+    /*
+    TODO: This test only works on 28+ due to a bug fix with the status
```
I'd like to leave the test implemented and just skip it.

I hate commented tests as it's harder to see them from the UI of the IDE and I tend to forget that we even have it

provokateurin replied:
Ah I forgot that you can skip tests

Leptopoda commented on packages/neon/neon/lib/src/models/account.dart:
```diff
@@ -41,6 +41,7 @@ class Account implements Credentials {
         );
 
   factory Account.fromJson(final Map<String, dynamic> json) => _$AccountFromJson(json);
+
```
Intened?

Leptopoda commented on packages/nextcloud/lib/src/helpers/news.dart:
```diff
@@ -0,0 +1,16 @@
+
+  const NewsListType(this.code);
+
+  final int code;
```
Is the code even needed?

`NewsListType.feed.index` should also yield the same (yes this requires correct ordering).

Leptopoda commented on packages/nextcloud/lib/src/helpers/core.dart:
```diff
@@ -0,0 +1,22 @@
+  usergroup(2),
+  link(3),
+  email(4),
+  // 5 was contact, is no longer used
```
I'd rather have the deprecated ones be implemented with a `deprecated` annotation.

That way `ShareType.values[index]` will always match the requested index.

Leptopoda commented on packages/nextcloud/lib/src/client.dart:
```diff
@@ -19,23 +21,31 @@ class NextcloudClient extends DynamiteClient {
-  })  : assert(loginName != null || username == null, 'Provide loginName instead of username or both'),
+  })  : assert(
+          ((loginName != null) == (username != null)) || (loginName != null || username == null),
+          'Provide loginName instead of username or both',
```
So `loginName` is always needed and `username` is optional?

In this case we should make the first one non nullable.

provokateurin replied:
The username variable was used for the WebDAV endpoint but that is no longer the case. It can just be removed. loginName still has to stay nullable because you can create clients without credentials.

provokateurin replied:
Fixed it in https://github.com/provokateurin/nextcloud-neon/pull/526

Leptopoda reviewed (commented):
Just reviewing the not generated code :sweat_smile: 

I'll still need to test it but it looks quite good.

Leptopoda commented:
I don't see any changes from my last review.
What should I review?

provokateurin commented:
Just approve :) Will push my rebase now

provokateurin commented:
Ah I forgot that I haven't pushed my changes yet :upside_down_face:

provokateurin commented:
@Leptopoda now you can review

Leptopoda reviewed (approved):
:tada:

provokateurin commented:
Actually to also close https://github.com/provokateurin/nextcloud-neon/issues/1 we need files_sharing

provokateurin commented:
Woop woop! A dream came true! :heart:
